### PR TITLE
[Refactor] FreeInit for AnimateDiff based pipelines

### PR DESCRIPTION
**Pull request by DN6**
# What does this PR do?
The current FreeInit implementation isn't super ideal. Based on the discussions in this PR https://github.com/huggingface/diffusers/pull/6644, 

Proposing this change based on the previous discussions that allows reusing the freeinit utils via a Mixin so that it is easy to experiment with adding the feature to other videos pipelines, while trying to avoid introducing two functions to run the denoising loop. This would preserve the existing pipeline denoising loop semantics so that the video pipelines can be read/understood like any other pipeline in `diffusers` and introducing FreeInit to a video pipeline does not involve adding a lot of boilerplate. 

This PR refactors FreeInit for 
AnimateDiff
PIA

And adds it to 
AnimateDiffVideotoVideo     

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by a-r-r-o-w — APPROVED** [review]
Looks great to me since we're mostly back to how this was first added! Also, maybe AnimateDiffControlnetPipeline in community pipelines could benefit from these changes as well wdyt?

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+    def enable_free_init(
+        self,
+        num_iters: int = 3,
+        use_fast_sampling: bool = False,
+        method: str = "butterworth",
+        order: int = 4,
+        spatial_stop_frequency: float = 0.25,
+        temporal_stop_frequency: float = 0.25,
+        generator: torch.Generator = None,
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/animatediff/freeinit_utils.py:86]: @DN6 I think it makes sense to get rid of the generator here since the one used in normal pipeline `__call__` can be passed. I made similar changes in #6644.
```suggestion
```

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                whereas lower values lead to `gaussian` method behaviour.
+            spatial_stop_frequency (`float`, *optional*, defaults to `0.25`):
+                Normalized stop frequency for spatial dimensions. Must be between 0 to 1. Referred to as `d_s` in
+                the original implementation.
+            temporal_stop_frequency (`float`, *optional*, defaults to `0.25`):
+                Normalized stop frequency for temporal dimensions. Must be between 0 to 1. Referred to as `d_t` in
+                the original implementation.
+            generator (`torch.Generator`, *optional*, defaults to `0.25`):
+                A [`torch.Generator`](https://pytorch.org/docs/stable/generated/torch.Generator.html) to make
+                FreeInit generation deterministic.
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/animatediff/freeinit_utils.py:112]:
```suggestion
```

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        """
+        self._free_init_num_iters = num_iters
+        self._free_init_use_fast_sampling = use_fast_sampling
+        self._free_init_method = method
+        self._free_init_order = order
+        self._free_init_spatial_stop_frequency = spatial_stop_frequency
+        self._free_init_temporal_stop_frequency = temporal_stop_frequency
+        self._free_init_generator = generator
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/animatediff/freeinit_utils.py:120]:
```suggestion
```

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            self.order,
+            self.spatial_stop_frequency,
+            self.temporal_stop_frequency,
+        )
+
+    def _apply_freq_filter(self, x: torch.Tensor, noise: torch.Tensor, low_pass_filter: torch.Tensor) -> torch.Tensor:
+        r"""Noise reinitialization."""
+        return _apply_freq_filter(x, noise, low_pass_filter)
+
+    def _apply_freeinit(self, latents, free_init_iteration, num_inference_steps, device, dtype):
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/animatediff/freeinit_utils.py:145]:
```suggestion
    def _apply_freeinit(self, latents, free_init_iteration, num_inference_steps, device, dtype, generator):
```

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        current_diffuse_timestep = self.scheduler.config.num_train_timesteps - 1
+        diffuse_timesteps = torch.full((latent_shape[0],), current_diffuse_timestep).long()
+
+        z_t = self.scheduler.add_noise(
+            original_samples=latents, noise=self._free_init_initial_noise, timesteps=diffuse_timesteps.to(device)
+        ).to(dtype=torch.float32)
+
+        z_rand = randn_tensor(
+            shape=latent_shape,
+            generator=self._free_init_generator,
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/animatediff/freeinit_utils.py:171]:
```suggestion
            generator=generator,
```

File: src/diffusers/pipelines/animatediff/pipeline_animatediff.py [hunk 6]
```diff
@@ -1060,43 +788,62 @@ def __call__(
...
-                device=device,
-            )
-        else:
-            latents = self._denoise_loop(**denoise_args)
-
-        video = self._retrieve_video_frames(latents, output_type, return_dict)
+        num_free_init_iters = self._free_init_num_iters if self.free_init_enabled else 1
+        for free_init_iter in range(num_free_init_iters):
+            if self.free_init_enabled:
+                latents = self._apply_freeinit(latents, free_init_iter, num_inference_steps, device, latents.dtype)
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/animatediff/pipeline_animatediff.py:794]:
```suggestion
                latents = self._apply_freeinit(latents, free_init_iter, num_inference_steps, device, latents.dtype, generator)
```

File: src/diffusers/pipelines/animatediff/pipeline_animatediff_video2video.py [hunk 7]
```diff
@@ -912,42 +915,48 @@ def __call__(
...
-
-                    latents = callback_outputs.pop("latents", latents)
-                    prompt_embeds = callback_outputs.pop("prompt_embeds", prompt_embeds)
-                    negative_prompt_embeds = callback_outputs.pop("negative_prompt_embeds", negative_prompt_embeds)
-
-                progress_bar.update()
+        num_free_init_iters = self._free_init_num_iters if self.free_init_enabled else 1
+        for free_init_iter in range(num_free_init_iters):
+            if self.free_init_enabled:
+                latents = self._apply_freeinit(latents, free_init_iter, num_inference_steps, device, latents.dtype)
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/animatediff/pipeline_animatediff_video2video.py:921]:
```suggestion
                latents = self._apply_freeinit(latents, free_init_iter, num_inference_steps, device, latents.dtype, generator)
```

File: src/diffusers/pipelines/pia/pipeline_pia.py [hunk 8]
```diff
@@ -1183,41 +1034,61 @@ def __call__(
...
-                device=device,
-            )
-        else:
-            latents = self._denoise_loop(**denoise_args)
-
-        video = self._retrieve_video_frames(latents, output_type, return_dict)
+        num_free_init_iters = self._free_init_num_iters if self.free_init_enabled else 1
+        for free_init_iter in range(num_free_init_iters):
+            if self.free_init_enabled:
+                latents = self._apply_freeinit(latents, free_init_iter, num_inference_steps, device, latents.dtype)
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pia/pipeline_pia.py:1040]:
```suggestion
                latents = self._apply_freeinit(latents, free_init_iter, num_inference_steps, device, latents.dtype, generator)
```

**Review by a-r-r-o-w — CHANGES_REQUESTED** [review]
Thanks!

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+    def _get_freeinit_freq_filter(self, shape: Tuple[int, ...]) -> torch.Tensor:
+        r"""Returns the FreeInit filter based on filter type and other input conditions."""
+        return _get_freeinit_freq_filter(
+            shape,
+            self.device,
+            self.filter_type,
+            self.order,
+            self.spatial_stop_frequency,
+            self.temporal_stop_frequency,
+        )
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/animatediff/freeinit_utils.py:139]: this is not used, no? we can just call `_get_freeinit_freq_filter` directly
```suggestion
```

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+        z_t = self.scheduler.add_noise(
+            original_samples=latents, noise=self._free_init_initial_noise, timesteps=diffuse_timesteps.to(device)
+        ).to(dtype=torch.float32)
+
+        z_rand = randn_tensor(
+            shape=latent_shape,
+            generator=self._free_init_generator,
+            device=device,
+            dtype=torch.float32,
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/animatediff/freeinit_utils.py:173]: any reason we use float32 here instead `dtype`?

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        )
+        latents = self._apply_freq_filter(z_t, z_rand, low_pass_filter=free_init_freq_filter)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/animatediff/freeinit_utils.py:175]:
```suggestion
        latents = _apply_freq_filter(z_t, z_rand, low_pass_filter=free_init_freq_filter)
```

**Review by yiyixuxu — COMMENTED** [review]
Thanks! I left some feedback :)
I'm ok with the nested loop here - it is pretty easy to understand now. 
The problem I can think of is if we want to incorporate other techniques that involve nested loops. But we can handle that later when we encounter it

File: src/diffusers/pipelines/animatediff/pipeline_animatediff.py [hunk 12]
```diff
@@ -535,63 +471,6 @@ def enable_freeu(self, s1: float, s2: float, b1: float, b2: float):
             raise ValueError("The pipeline must have `unet` for using FreeU.")
         self.unet.enable_freeu(s1=s1, s2=s2, b1=b1, b2=b2)
 
-    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.disable_freeu
```
[the remark this answers is not in the document]
Comment (DN6) [reply src/diffusers/pipelines/animatediff/pipeline_animatediff.py:538]: Whoops. Mistake here. Thanks for catching.

File: src/diffusers/pipelines/animatediff/pipeline_animatediff.py (hunk 6 above)
Comment (patrickvonplaten) [inline src/diffusers/pipelines/animatediff/pipeline_animatediff.py:794]:
```suggestion
                latents = self._apply_free_init(latents, free_init_iter, num_inference_steps, device, latents.dtype)
```

More consistent

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
+import math
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/pipelines/animatediff/freeinit_utils.py:1]: Also add license here on top. Agree with Yiyi, let's maybe move all of this into a "utils" file?

**Review by patrickvonplaten — COMMENTED** [review]
Very nice refactor overall!

File: src/diffusers/pipelines/animatediff/pipeline_animatediff.py [hunk 14]
```diff
@@ -1060,43 +788,62 @@ def __call__(
...
-                batch_size=batch_size,
-                num_videos_per_prompt=num_videos_per_prompt,
-                denoise_args=denoise_args,
-                device=device,
-            )
-        else:
-            latents = self._denoise_loop(**denoise_args)
-
-        video = self._retrieve_video_frames(latents, output_type, return_dict)
+        num_free_init_iters = self._free_init_num_iters if self.free_init_enabled else 1
```
[the remark this answers is not in the document]
Comment (DN6) [reply src/diffusers/pipelines/animatediff/pipeline_animatediff.py:802]: I thought about it, but since we're always setting `self._free_init_iters` even if `free_init` isn't enabled, we would get an extra progress bar all the time. We could introduce additional checks to display the progress bar only when free init is enabled but it might not be worth it IMO.

File: src/diffusers/pipelines/animatediff/freeinit_utils.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+        return _get_freeinit_freq_filter(
+            shape,
+            self.device,
+            self.filter_type,
+            self.order,
+            self.spatial_stop_frequency,
+            self.temporal_stop_frequency,
+        )
+
+    def _apply_freq_filter(self, x: torch.Tensor, noise: torch.Tensor, low_pass_filter: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (DN6) [reply src/diffusers/pipelines/animatediff/freeinit_utils.py]: I moved these functions into the FreeInitMixin. I don't think they really need to be accessed outside the Mixin.

**Review by yiyixuxu — APPROVED** [review]
thanks!
do we need a FreeInitTexterMixin? If it makes sense we can put out an issue for the community to take up; But if it's just going to be added to 4 or 5 pipelines might not needed

**Comment by DN6** [comment]
@yiyixuxu I think we limit it to just the AnimateDiff based pipelines for now until we see some usage/requests from the community to add to other pipelines. We can add the testing mixin if this expands to >5 pipelines?

**Comment by a-r-r-o-w** [comment]
@DN6 Btw does it make sense to also add to [AnimateDiff ControlNet](https://github.com/huggingface/diffusers/blob/main/examples/community/pipeline_animatediff_controlnet.py) (community)?

**Comment by DN6** [comment]
@a-r-r-o-w Yeah sure thing 👍🏽. 

BTW @a-r-r-o-w AnimateLCM checkpoints are available in diffusers format now. 
https://huggingface.co/wangfuyun/AnimateLCM

I tried running FreeInit with AnimateLCM and I ended up with a bunch of noise in the final output. If you're up to it, would you like to run some tests here to pin point the issue? I suspect it has to do with the LCM Scheduler, but don't have the time to really dig into it. LCM + FreeInit could be very nice for better quality videos.

File: src/diffusers/pipelines/free_init_utils.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+        latents: torch.Tensor,
+        free_init_iteration: int,
+        num_inference_steps: int,
+        device: torch.device,
+        dtype: torch.dtype,
+        generator: torch.Generator,
+    ):
+        if free_init_iteration == 0:
+            self._free_init_initial_noise = latents.detach().clone()
+            return latents, self.scheduler.timesteps
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/free_init_utils.py:149]: @DN6 This is incorrect and seems like a regression from the old implementation. I was trying to debug why AnimateLCM was failing to produce good results and stumbled upon this other issue (it does produce good results btw except when for use_fast_sampling==False. setting it to True seems to give good results).

- Copy FreeInit code from [here](https://huggingface.co/docs/diffusers/v0.26.3/en/api/pipelines/animatediff#using-freeinit). You will see that the first iteration runs for 20 steps, second iteration runs for 13 steps and third iteration runs for 20 steps. This is incorrect because when `use_fast_sampling=True`, it should be 7, 13 and 20 but we return here without the fast sampling check.

**Comment by a-r-r-o-w** [comment]
> @a-r-r-o-w Yeah sure thing 👍🏽.
> 
> BTW @a-r-r-o-w AnimateLCM checkpoints are available in diffusers format now. https://huggingface.co/wangfuyun/AnimateLCM
> 
> I tried running FreeInit with AnimateLCM and I ended up with a bunch of noise in the final output. If you're up to it, would you like to run some tests here to pin point the issue? I suspect it has to do with the LCM Scheduler, but don't have the time to really dig into it. LCM + FreeInit could be very nice for better quality videos.

After some testing, it seems like when one uses `use_fast_sampling=False`, the results are all noisy no matter what the underlying freeinit method is. Setting it to True always gives me good results. Note that the results below have been done after fixing the issue mentioned [here](https://github.com/huggingface/diffusers/pull/6874#discussion_r1493756055).

<details>
<summary> Code </summary>

```py
import torch
from diffusers import MotionAdapter, AnimateDiffPipeline, LCMScheduler
from diffusers.utils import export_to_gif

adapter = MotionAdapter.from_pretrained("wangfuyun/AnimateLCM")
model_id = "SG161222/Realistic_Vision_V5.1_noVAE"
pipe = AnimateDiffPipeline.from_pretrained(model_id, motion_adapter=adapter, torch_dtype=torch.float16).to("cuda")
pipe.scheduler = LCMScheduler.from_pretrained(
    model_id,
    subfolder="scheduler",
    beta_schedule="linear",
    clip_sample=False,
    timestep_spacing="linspace",
    steps_offset=1
)

pipe.load_lora_weights("wangfuyun/AnimateLCM", weight_name="sd15_lora_beta.safetensors", adapter_name="lcm-lora")
pipe.set_adapters(["lcm-lora"], [0.8])

pipe.enable_vae_slicing()
pipe.enable_free_init(num_iters=3, method="gaussian", use_fast_sampling=True)

output = pipe(
    prompt="a panda playing a guitar, on a boat, in the ocean, high quality",
    negative_prompt="bad quality, worse quality",
    num_frames=16,
    guidance_scale=2.5,
    num_inference_steps=6,
    generator=torch.Generator("cpu").manual_seed(666),
)
frames = output.frames[0]
export_to_gif(frames, "animation.gif")
```
</details>

<table>
<tr>
<td></td>
<td></td>
<td></td>
</tr>
</table>

File: src/diffusers/pipelines/free_init_utils.py (hunk 16 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/free_init_utils.py:149]: @DN6 Could I open a PR fixing this behavior since this has been merged already?

Comment (DN6) [reply src/diffusers/pipelines/free_init_utils.py:149]: Hi @a-r-r-o-w missed this. Yes please feel free to open a PR.
